### PR TITLE
Introduce JumpdestRemover optimisation step

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
Some of the bigger assemblies in the Parity multisig (assembly items: before, after):
```
5474 -> 5430
3857 -> 3802
3638 -> 3635
3616 -> 3616
```

The function dispatcher leaves a lot of stray jumpdests, we could fix that or could keep a simple remover such as this.

**Comment by axic** [comment]
Haven't bothered updating the tests which check the exact bytecode before the code is accepted.

**Comment by axic** [comment]
Probably this should be run again after dedup/CSE.

**Comment by chriseth** [comment]
This step was previously done by the control flow graph optimizer, but it was removed because we currently cannot assume that labels that are never pushed can be removed, because they can be stored in storage. If we do a full analysis also passing in referenced labels "from above", it might work. In any case, there should be a test that fails if you do not consider this.

**Comment by axic** [comment]
> we currently cannot assume that labels that are never pushed can be removed, because they can be stored in storage

Hm, yes that's true, this change would break that.

Thinking a bit more about that, is that valid usage though? If it is at any point of time pushed to storage in the code/constructor, it is already safe.

In an example like ```{ 4 calldataload jump } ```, where the tag is received from the calldata it wouldn't work. The question is: is that a valid use case we want to support? I don't think so.

**Comment by chriseth** [comment]
When the jumpdest remover as currently implemented analyses the runtime assembly, it does not consider the jumpdests from the runtime assembly that were accessed in the creation time assembly.

**Comment by axic** [comment]
> that were accessed in the creation time assembly.

True. Is that an actual case we have? I guess if helpers are used in a constructor which are also used  in runtime functions. Strange no tests caught this, should add a test for it.

**Comment by chriseth** [comment]
The test `store_internal_unused_function_in_constructor` should fail.

**Comment by axic** [comment]
Doesn't seem to fail on that one.

**Comment by chriseth** [comment]
As can be seen here - https://medium.com/@hayeah/diving-into-the-ethereum-vm-part-2-storage-layout-bc5349cb11b7 - it seems we have a lot of jumpdests that hinder peephole or CSE optimization, so increasing the priority of this.

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -456,6 +465,14 @@ map<u256, u256> Assembly::optimiseInternal(OptimiserSettings _settings)
 		}
 	}
 
+	/// Run again after Deduplicate or CSE
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]: why is this needed? We will just have another iteration of the loop, won't we?

**Comment by chriseth** [comment]
Ok, now I'm relieved: The test does actually fail.

File: libevmasm/Assembly.cpp (hunk 1 above)
Comment (axic) [reply libevmasm/Assembly.cpp]: I've seen some cases where dedup/cse left unused tags. Admittedly they could be improved to not do that.

**Comment by axic** [comment]
Can you add some specific tests (including subassemblies) to `test/libevmasm/Optimiser`?

**Comment by chriseth** [comment]
Added test.

File: libevmasm/JumpdestRemover.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	/// @returns a set of all tags from the given sub-assembly that are referenced
+	/// from the given list of items.
+	static std::set<size_t> referencedTags(AssemblyItems const& _items, size_t _subId);
+
+private:
+	void analyzeIfNecessary() const;
+
+	AssemblyItems& m_items;
+	bool mutable m_initialized = false;
+	std::set<std::pair<size_t, size_t>> mutable m_referencedTags;
```
Comment (axic) [inline libevmasm/JumpdestRemover.h]: These two are unused.

File: libevmasm/JumpdestRemover.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	explicit JumpdestRemover(AssemblyItems& _items): m_items(_items) {}
+
+	bool optimise(std::set<size_t> const& _tagsReferencedFromOutside);
+
+	/// @returns a set of all tags from the given sub-assembly that are referenced
+	/// from the given list of items.
+	static std::set<size_t> referencedTags(AssemblyItems const& _items, size_t _subId);
+
+private:
+	void analyzeIfNecessary() const;
```
Comment (axic) [inline libevmasm/JumpdestRemover.h]: Not used.

**Comment by chriseth** [comment]
Weird that the compiler does not warn about that...

**Comment by chriseth** [comment]
This is ready to be merged.

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
@chriseth as discussed last meeting we wanted to run the truffle tests on openzeppelin with this change.

**Comment by chriseth** [comment]
What about merging and then testing the nightly?

**Comment by axic** [comment]
Depends on #2782.

**Comment by axic** [comment]
Rebased, merging this next.
